tests/Temporalio.Tests/Worker/ActivityWorkerTests.cs

```diff
@@ -0,0 +1,718 @@
+    }
+
+    [Fact]
+    public async Task ExecuteActivityAsync_CaughtCancel_Succeeds()
+    {
+        var activityReached = new TaskCompletionSource();
+        [Activity]
+        async Task<string> CatchCancelledAsync()
+        {
+            activityReached.SetResult();
+            while (!ActivityContext.Current.CancellationToken.IsCancellationRequested)
+            {
+                ActivityContext.Current.Heartbeat();
+                await Task.Delay(300);
+            }
+            return "Cancelled!";
+        }
+        var res = await ExecuteActivityAsync(
+            CatchCancelledAsync,
+            waitForCancellation: true,
+            heartbeatTimeout: TimeSpan.FromSeconds(1),
+            afterStarted: async handle =>
+            {
+                // Wait for activity to be reached, then cancel the workflow
+                await activityReached.Task.WaitAsync(TimeSpan.FromSeconds(20));
+                await handle.CancelAsync();
+            });
+        Assert.Equal("Cancelled!", res);
+    }
+
+    [Fact]
+    public async Task ExecuteActivityAsync_WorkerShutdown_ReportsFailure()
+    {
+        using var workerStoppingSource = new CancellationTokenSource();
+        var activityReached = new TaskCompletionSource();
+        var gotCancellation = false;
+        [Activity]
+        async Task WaitUntilCancelledAsync()
+        {
+            activityReached.SetResult();
+            while (!ActivityContext.Current.CancellationToken.IsCancellationRequested)
+            {
+                ActivityContext.Current.Heartbeat();
+                await Task.Delay(300);
+            }
+            gotCancellation = true;
+            ActivityContext.Current.CancellationToken.ThrowIfCancellationRequested();
+        }
+        var workflowID = string.Empty;
+        await Assert.ThrowsAnyAsync<OperationCanceledException>(() => ExecuteActivityAsync(
+            WaitUntilCancelledAsync,
+            workerStoppingToken: workerStoppingSource.Token,
+            afterStarted: async handle =>
+            {
+                workflowID = handle.ID;
+                // Wait for activity to be reached, then stop the worker
+                await activityReached.Task.WaitAsync(TimeSpan.FromSeconds(20));
+                workerStoppingSource.Cancel();
+            }));
+        Assert.True(gotCancellation);
+        // Check the workflow error
+        var wfErr = await Assert.ThrowsAsync<WorkflowFailedException>(
+            () => Client.GetWorkflowHandle(workflowID).GetResultAsync());
+        var actErr = Assert.IsType<ActivityFailureException>(wfErr.InnerException);
+        Assert.IsType<ApplicationFailureException>(actErr.InnerException);
+    }
+
+    [Fact]
+    public async Task ExecuteActivityAsync_ThrowsOperationCancelled_ReportsFailure()
+    {
+        // Just to confirm there's not an issue with this exception
+        [Activity]
+        static void Throws() => throw new OperationCanceledException();
```

Comment:
Why is this test required, isn't it a duplicate of `ExecuteActivityAsync_ThrowCommonException_ReportsFailure`?
Are we testing what happens when there's no error message?

Reply:
From the comment:

> // Just to confirm there's not an issue with this exception

In Python I had an issue where I caught the language native cancelled on a non cancelling scenario and tried to send cancel back to server only to have server fail saying cancel was never requested. This is just a sanity test to confirm I don't have that mistake here.

Reply:
Okay, maybe worth a more detailed comment if you want.